File: proposed-projects/round-trip-experiment/reverse/20260909T184722Z_089a4294/20260910T101353Z_4b392f72/ported_implementation/src/utils/errors/build-error-position.ts
export const MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW = 3;

export const buildErrorPosition = (src: string, pos: number): string[] => {
  if (src === '') {
    return ['No input provided'];
  }

  const lines = src.split('\n');

  let lineIdx = 0;
  while (lineIdx < lines.length && lines[lineIdx] && pos > lines[lineIdx].length - 1) {
    pos -= lines[lineIdx].length;
    lineIdx += 1;
  }

  const linesToShow: string[] = [];
  const firstLine = Math.max(0, lineIdx - (MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW - 1));
  for (let i = firstLine; i <= lineIdx; i++) {
    // joining the output turns an out-of-bounds line into an empty string.
    linesToShow.push(i < lines.length ? lines[i] : '');
  }

  return [...linesToShow, `${' '.repeat(Math.max(0, pos))}^`];
};

File: packages/gbnf-experiment/data/20260909T184722Z_089a4294/ported_implementation/gbnf/utils/errors/build_error_position.py
from __future__ import annotations

MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW = 3


def build_error_position(src: str, pos: int) -> list[str]:
    if src == "":
        return ["No input provided"]

    lines = src.split("\n")

    line_idx = 0
    while line_idx < len(lines) and lines[line_idx] and pos > len(lines[line_idx]) - 1:
        pos -= len(lines[line_idx])
        line_idx += 1

    lines_to_show: list[str] = []
    for i in range(max(0, line_idx - (MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW - 1)), line_idx + 1):
        # the reference implementation joins its output, which turns an out-of-bounds
        # line into an empty string.
        lines_to_show.append(lines[i] if i < len(lines) else "")

    return [*lines_to_show, " " * max(0, pos) + "^"]
